DietPi-Config | Display Brightness Changes

Fourdee opened the pull request:
REF: https://github.com/MichaIng/DietPi/issues/6284 

Following changes:
- Adds support for setting multiple backlight displays
- Values are now in a human readable format (%)
- OLED sysfs support (and DietPi) requires kernel 5.12 or higher.

Notes:
- Backlight brightness values are saved during reboot (confirmed with ```intel_backlight```). So no SystemD or additional boot calls required.
- Give me some time to make Shellcheck happy :smile: 

---

New menu to list all available backlights:
![image](https://user-images.githubusercontent.com/7077026/229808505-9a889be8-448c-4c82-b259-b624a59e3fd2.png)

New select % menu:
![image](https://user-images.githubusercontent.com/7077026/229808715-136e1b9d-41ec-46be-917e-f4eddd721312.png)

Fourdee committed:
DietPi-Config | Display Brightness

Support multiple backlights in menu.
Convert all brightness values to % for human readable format.

Fourdee committed:
minor

Remove debugging

Fourdee committed:
Tweaks

- Switch from hard coded backlight locations, to a find system ```/sys/class/backlight/*```
- Switch from array index in menus, to backlight name during selection.

Fourdee committed:
Correction

- For loop.

MichaIng commented on dietpi/dietpi-config:
```diff
@@ -430,32 +430,75 @@ A long (or insufficiently manufactured) cable may required a higher boost settin
-				if G_WHIP_INPUTBOX "Please enter a brightness value:\n - Min = $MIN_VALUE | Max = $MAX_VALUE" && G_CHECK_VALIDINT "$G_WHIP_RETURNED_VALUE" "$MIN_VALUE" "$MAX_VALUE"; then
+					aBrightness_Percent+=( $(echo ${aBrightness_Current[i]} ${aBrightness_Max[i]} | awk '{printf "%.0f\n", $1 / $2 * 100}') )
+					#G_WHIP_MENU_ARRAY+=("$i" " : ${aBrightness_Percent[i]}% [${aBrightness_Name[i]}]" )
+					G_WHIP_MENU_ARRAY+=("${aBrightness_Name[i]}" " : [${aBrightness_Percent[i]}%]" )
```
```suggestion
					G_WHIP_MENU_ARRAY+=("${aBrightness_Name[$i]}" " : [${aBrightness_Percent[$i]}%]" )
```

MichaIng commented on dietpi/dietpi-config:
```diff
@@ -430,32 +430,75 @@ A long (or insufficiently manufactured) cable may required a higher boost settin
+					# - Match G_WHIP_RETURNED_VALUE to array index
+					for (( i=0; i<$iTotalAvailable; i++ ))
+					do
+						if [[ ${aBrightness_Name[i]} == $G_WHIP_RETURNED_VALUE ]]
```
```suggestion
						if [[ ${aBrightness_Name[$i]} == $G_WHIP_RETURNED_VALUE ]]
```

MichaIng commented on dietpi/dietpi-config:
```diff
@@ -430,32 +430,75 @@ A long (or insufficiently manufactured) cable may required a higher boost settin
+						fi
+					done
+
+					G_WHIP_DEFAULT_ITEM=${aBrightness_Percent[iIndex]}
```
```suggestion
					G_WHIP_DEFAULT_ITEM=${aBrightness_Percent[$iIndex]}
```

MichaIng commented on dietpi/dietpi-config:
```diff
@@ -430,32 +430,75 @@ A long (or insufficiently manufactured) cable may required a higher boost settin
+                	                        G_WHIP_MENU_ARRAY+=("$(( $i * 10 ))" '%')
+                        	        done
+
+	                                if G_WHIP_MENU "Please select a new Brightness level for [${aBrightness_Name[iIndex]}]:\n - Current Value = ${aBrightness_Percent[iIndex]}%"
```
```suggestion
	                                if G_WHIP_MENU "Please select a new Brightness level for [${aBrightness_Name[$iIndex]}]:\n - Current Value = ${aBrightness_Percent[$iIndex]}%"
```

MichaIng commented on dietpi/dietpi-config:
```diff
@@ -430,32 +430,75 @@ A long (or insufficiently manufactured) cable may required a higher boost settin
+	                                if G_WHIP_MENU "Please select a new Brightness level for [${aBrightness_Name[iIndex]}]:\n - Current Value = ${aBrightness_Percent[iIndex]}%"
+					then
+
+	        	                        local iBrightness_Value_New=$(( ( ${aBrightness_Max[iIndex]} / 100 ) * $G_WHIP_RETURNED_VALUE ))
```
```suggestion
	        	                        local iBrightness_Value_New=$(( ( ${aBrightness_Max[$iIndex]} / 100 ) * $G_WHIP_RETURNED_VALUE ))
```

MichaIng commented on dietpi/dietpi-config:
```diff
@@ -430,32 +430,75 @@ A long (or insufficiently manufactured) cable may required a higher boost settin
+					then
+
+	        	                        local iBrightness_Value_New=$(( ( ${aBrightness_Max[iIndex]} / 100 ) * $G_WHIP_RETURNED_VALUE ))
+                	        	        echo $iBrightness_Value_New > "${aFp_Brightness_Available[iIndex]}/brightness"
```
```suggestion
                	        	        echo $iBrightness_Value_New > "${aFp_Brightness_Available[$iIndex]}/brightness"
```

MichaIng commented on dietpi/dietpi-config:
```diff
@@ -430,32 +430,75 @@ A long (or insufficiently manufactured) cable may required a higher boost settin
 
-				G_WHIP_DEFAULT_ITEM=$current_brightness
-				if G_WHIP_INPUTBOX "Please enter a brightness value:\n - Min = $MIN_VALUE | Max = $MAX_VALUE" && G_CHECK_VALIDINT "$G_WHIP_RETURNED_VALUE" "$MIN_VALUE" "$MAX_VALUE"; then
+					aBrightness_Percent+=( $(echo ${aBrightness_Current[i]} ${aBrightness_Max[i]} | awk '{printf "%.0f\n", $1 / $2 * 100}') )
```
```suggestion
					aBrightness_Percent+=( $(echo ${aBrightness_Current[$i]} ${aBrightness_Max[$i]} | awk '{printf "%.0f\n", $1 / $2 * 100}') )
```

Fourdee committed:
Shellcheck

- Shellcheck sweeties :candy: 
- Also check ```max_brightness``` valid int

MichaIng commented on dietpi/dietpi-config:
```diff
@@ -430,32 +430,76 @@ A long (or insufficiently manufactured) cable may required a higher boost settin
 
-				G_WHIP_DEFAULT_ITEM=$current_brightness
-				if G_WHIP_INPUTBOX "Please enter a brightness value:\n - Min = $MIN_VALUE | Max = $MAX_VALUE" && G_CHECK_VALIDINT "$G_WHIP_RETURNED_VALUE" "$MIN_VALUE" "$MAX_VALUE"; then
+					aBrightness_Percent+=( "$(echo ${aBrightness_Current[i]} ${aBrightness_Max[i]} | awk '{printf "%.0f\n", $1 / $2 * 100}')" )
```
Indeed within command substitutions `$(...)`, everything needs to be quoted again. Otherwise unintended globbing and word splitting can happen within the command substitution and lead to an unexpected result before the parent quotes take effect.
```suggestion
					aBrightness_Percent+=( "$(echo "${aBrightness_Current[i]} ${aBrightness_Max[i]}" | mawk '{printf "%.0f\n", $1 / $2 * 100}')" )
```
I also switched from `awk` to `mawk`, which is much faster and supports all we need.

MichaIng commented on dietpi/dietpi-config:
```diff
@@ -430,32 +430,76 @@ A long (or insufficiently manufactured) cable may required a higher boost settin
+					aFp_Brightness_Available+=( "$fp_brightness" )
+					aBrightness_Name+=( "${fp_brightness##*/}" )
+					aBrightness_Current+=( "$(<$fp_brightness/brightness)" )
+					aBrightness_Max+=( "$(<$fp_brightness/max_brightness)" )
```
Indeed that's it, also command substitutions, as `<$fp_brightness/brightness` would be a syntax error if `$fp_brightness` contained a space and it if it did contain a special character like `[]` or `*` and by chance a matching file exists, it can be glob-expanded:
```suggestion
					aBrightness_Current+=( "$(<"$fp_brightness/brightness")" )
					aBrightness_Max+=( "$(<"$fp_brightness/max_brightness")" )
```

Fourdee committed:
Update dietpi/dietpi-config

Thanks Micha

Co-authored-by: MichaIng <micha@dietpi.com>

Fourdee committed:
Update dietpi/dietpi-config

Thanks Micha

Co-authored-by: MichaIng <micha@dietpi.com>

Fourdee committed:
DietPi-Config: Display Brightness

https://github.com/MichaIng/DietPi/pull/6288

Fourdee committed:
Display Brightness Changes

- :u6307: Use higher precision floating point MAWK during apply value process (from % to raw value)
- :u6307: Only show multiple selection menu, if multiple options exist
- :u6307: Add custom option input box 
- :u6307: Use ```actual_brightness``` for reading current values as reported by HW, as per:

Fourdee committed:
Shellcheck

- Line space removal

Fourdee committed:
Shellcheck

- Optimize find and replace char in string

MichaIng commented on dietpi/dietpi-config:
```diff
@@ -430,33 +430,95 @@ A long (or insufficiently manufactured) cable may required a higher boost settin
+					aBrightness_Current+=( "$(<"$fp_brightness/actual_brightness")" )
+					aBrightness_Max+=( "$(<"$fp_brightness/max_brightness")" )
+					(( iTotalAvailable++ ))
+				fi
```
I think I'd take the risk and for now trust in that if a device is registered as backlight device, the sysfs API is complete. It doesn't seem to be an API which changed recently:
```suggestion
				aFp_Brightness_Available+=( "$fp_brightness" )
				aBrightness_Name+=( "${fp_brightness##*/}" )
				aBrightness_Current+=( "$(<"$fp_brightness/actual_brightness")" )
				aBrightness_Max+=( "$(<"$fp_brightness/max_brightness")" )
				(( iTotalAvailable++ ))
```

MichaIng commented on dietpi/dietpi-config:
```diff
@@ -430,33 +430,95 @@ A long (or insufficiently manufactured) cable may required a higher boost settin
-					echo "$G_WHIP_RETURNED_VALUE" > "$fp_brightness/brightness"
+					done
+					# - Match G_WHIP_RETURNED_VALUE to array index
+					i=0 #NB: Declared locally above.
```
Redundant as `for` loops set it forcefully anyway.
```suggestion
```

MichaIng commented on dietpi/dietpi-config:
```diff
@@ -430,33 +430,95 @@ A long (or insufficiently manufactured) cable may required a higher boost settin
+					G_WHIP_MENU_ARRAY=('Custom' ' : Manually enter a value')
+					# - Generate % menu
+					MIN_VALUE=0 MAX_VALUE=100
+					i=0 #NB: Declared locally above.
```
```suggestion
```

MichaIng reviewed (dismissed):
I've not tested it yet, but looks great. Only minor coding suggestions from my end.

MichaIng committed:
Merge branch 'dev' into dev

MichaIng committed:
Update dietpi/dietpi-config

MichaIng committed:
Update dietpi/dietpi-config

MichaIng committed:
Update dietpi/dietpi-config
